File: app/src/test/java/com/readtracker/android/db/export/JSONExporterTest.java
package com.readtracker.android.db.export;

import com.readtracker.android.db.Book;
import com.readtracker.android.db.Quote;
import com.readtracker.android.db.Session;
import com.readtracker.android.support.Utils;
import com.readtracker.android.test_support.DatabaseTestCase;
import com.readtracker.android.test_support.TestUtils;

import org.json.JSONObject;
import org.junit.Ignore;
import org.junit.Test;
import org.skyscreamer.jsonassert.JSONAssert;
import org.skyscreamer.jsonassert.JSONCompareMode;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.Arrays;
import java.util.List;

import static com.readtracker.android.test_support.TestUtils.buildSession;

public class JSONExporterTest extends DatabaseTestCase {

  private JSONExporter getExporter() {
    return new JSONExporter(getDatabaseManager());
  }

  @Test
  public void jsonExporterTest_ExportPopulatedBook_ReturnsStringOfOutputPath() throws Exception {
    List<Book> books = populateBooksForExpectedOutput();

    // Convert to string to get type agnostic comparison
    String actual = getExporter().exportAll(books).toString();
    String expected = TestUtils.readFixtureFile("expected_output_of_populated_book_test.json");

    JSONAssert.assertEquals(expected, actual, true);
  }

  @Test
  // Similar to export_all_books, but tests the write out to disk
  public void jsonExporterTest_ExportPopulatedBook_ReturnsExportedJsonFromIO() throws Exception {
    List<Book> books = populateBooksForExpectedOutput();

    String exportFilename = TestUtils.randomString();
    File exportFile = new File(getContext().getFilesDir(), exportFilename);

    getExporter().exportToDisk(books, exportFile);

    InputStream inputStream = new FileInputStream(getContext().getFileStreamPath(exportFilename));
    String exportFileContent = Utils.readInputStream(inputStream);

    JSONObject actual = new JSONObject(exportFileContent);
    String expected = TestUtils.readFixtureFile("expected_output_of_populated_book_test.json");

    JSONAssert.assertEquals(expected, actual, JSONCompareMode.NON_EXTENSIBLE);
  }

  @Ignore
  private List<Book> populateBooksForExpectedOutput() {
    Book metamorphosis = new Book();
    metamorphosis.setTitle("Metamorphosis");
    metamorphosis.setAuthor("Franz Kafka");
    metamorphosis.setClosingRemark("Dude, poor guy!");
    metamorphosis.setCoverImageUrl("https://images-of-books/metamorphosis.png");
    metamorphosis.setCurrentPosition(0.5f);
    metamorphosis.setCurrentPositionTimestampMs(213456789L);
    metamorphosis.setFirstPositionTimestampMs(123456789L);
    metamorphosis.setPageCount(123.45f);
    metamorphosis.setState(Book.State.Finished);
    getDatabaseManager().save(metamorphosis);

    Quote unicornQuote = TestUtils.buildQuote(metamorphosis, "unicorn", 0.45f, 123456799L);
    metamorphosis.getQuotes().add(unicornQuote);
    getDatabaseManager().save(unicornQuote);

    Quote guppyQuote = TestUtils.buildQuote(metamorphosis, "guppy", 0f, 4564321L);
    metamorphosis.getQuotes().add(guppyQuote);
    getDatabaseManager().save(guppyQuote);

    Session firstSession = buildSession(metamorphosis, 0.45f, 0.75f, 1234, 4000000L);
    metamorphosis.getSessions().add(firstSession);
    getDatabaseManager().save(firstSession);

    Session secondSession = buildSession(metamorphosis, 0.75f, 1.0f, 600, 5000000L);
    metamorphosis.getSessions().add(secondSession);
    getDatabaseManager().save(secondSession);

    Book androidFurDummies = new Book();
    androidFurDummies.setTitle("Android Apps Entwicklung für Dummies");
    androidFurDummies.setAuthor("Donn Felker");
    androidFurDummies.setState(Book.State.Finished);
    androidFurDummies.setCurrentPosition(0.00872093066573143f);
    androidFurDummies.setCurrentPositionTimestampMs(1400856553800L);
    androidFurDummies.setPageCount(344f);
    androidFurDummies.setCoverImageUrl("http://bks8.books.google.de/books?id=KPjmuog");
    getDatabaseManager().save(androidFurDummies);

    {
      Session session = buildSession(androidFurDummies, 0f, 0.00872093066573143f, 3, 1399535743000L);
      androidFurDummies.getSessions().add(session);
      getDatabaseManager().save(session);
    }

    {
      Session session = buildSession(androidFurDummies, 0.008720931f, 0.9800000190734863f, 1146, 1400856553800L);
      androidFurDummies.getSessions().add(session);
      getDatabaseManager().save(session);
    }

    Book emptyBook = new Book();
    getDatabaseManager().save(emptyBook);

    return Arrays.asList(metamorphosis, androidFurDummies, emptyBook);
  }
}

File: app/src/test/java/com/readtracker/android/db/export/JSONExporterTest.kt
package com.readtracker.android.db.export

import android.support.test.filters.SmallTest
import com.readtracker.android.db.Book
import com.readtracker.android.support.DatabaseTestCase
import com.readtracker.android.support.TestUtils
import com.readtracker.android.support.TestUtils.buildSession
import com.readtracker.android.support.Utils
import org.json.JSONObject
import org.junit.Test
import org.skyscreamer.jsonassert.JSONAssert
import org.skyscreamer.jsonassert.JSONCompareMode
import java.io.File
import java.io.FileInputStream
import java.util.*

class JSONExporterTest : DatabaseTestCase() {

    /**
     * Helper method for returning the JSON serializer
     * @return JSONExporter
     */
    private val exporter: JSONExporter
        get() = JSONExporter(databaseManager)

    /**
     * Get a list of books that we assemble, export it into
     * JSON and assert agains the expected output.
     * @throws Exception
     */
    @Test
    @SmallTest
    @Throws(Exception::class)
    fun jsonExporterTest_ExportPopulatedBook_ReturnsStringOfOutputPath() {
        val books = populateBooksForExpectedOutput()

        // Convert to string to get type agnostic comparison
        val actual = exporter.exportAll(books).toString()
        val expected = TestUtils.readFixtureFile("expected_output_of_populated_book_test.json")

        JSONAssert.assertEquals(expected, actual, true)
    }

    /**
     * Similar to [jsonExporterTest_ExportPopulatedBook_ReturnsStringOfOutputPath],
     * but tests the write out to disk.
     * @throws Exception
     */
    @Test
    @SmallTest
    @Throws(Exception::class)
    fun jsonExporterTest_ExportPopulatedBook_ReturnsExportedJsonFromIO() {
        val books = populateBooksForExpectedOutput()

        val exportFilename = TestUtils.randomString()
        val exportFile = File(getContext().filesDir, exportFilename)

        exporter.exportToDisk(books, exportFile)

        val inputStream = FileInputStream(getContext().getFileStreamPath(exportFilename))
        val exportFileContent = Utils.readInputStream(inputStream)

        val actual = JSONObject(exportFileContent)
        val expected = TestUtils.readFixtureFile("expected_output_of_populated_book_test.json")

        JSONAssert.assertEquals(expected, actual, JSONCompareMode.NON_EXTENSIBLE)
    }

    /**
     * Helper method to assemble a list of [Book]s with known data.
     * @return [List] of [Book] entities with data, [Quote]s and [Session]s.
    */
    private fun populateBooksForExpectedOutput(): List<Book> {
        val metamorphosis = Book().apply {
            title = "Metamorphosis"
            author = "Franz Kafka"
            closingRemark = "Dude, poor guy!"
            coverImageUrl = "https://images-of-books/metamorphosis.png"
            currentPosition = 0.5f
            currentPositionTimestampMs = 213456789L
            firstPositionTimestampMs = 123456789L
            pageCount = 123.45f
            state = Book.State.Finished
        }
        databaseManager.save(metamorphosis)

        TestUtils.buildQuote(metamorphosis, "unicorn", 0.45f, 123456799L).let { unicornQuote ->
            metamorphosis.quotes.add(unicornQuote)
            databaseManager.save(unicornQuote)
        }

        TestUtils.buildQuote(metamorphosis, "guppy", 0f, 4564321L).let { guppyQuote ->
            metamorphosis.quotes.add(guppyQuote)
            databaseManager.save(guppyQuote)
        }

        buildSession(metamorphosis, 0.45f, 0.75f, 1234, 4000000L).let { firstSession ->
            metamorphosis.sessions.add(firstSession)
            databaseManager.save(firstSession)
        }

        buildSession(metamorphosis, 0.75f, 1.0f, 600, 5000000L).let { secondSession ->
            metamorphosis.sessions.add(secondSession)
            databaseManager.save(secondSession)
        }

        val androidFurDummies = Book().apply {
            title = "Android Apps Entwicklung für Dummies"
            author = "Donn Felker"
            state = Book.State.Finished
            currentPosition = 0.00872093066573143f
            currentPositionTimestampMs = 1400856553800L
            pageCount = 344f
            coverImageUrl = "http://bks8.books.google.de/books?id=KPjmuog"
        }

        databaseManager.save(androidFurDummies)

        buildSession(androidFurDummies, 0f, 0.00872093066573143f, 3, 1399535743000L).let { session ->
            androidFurDummies.sessions.add(session)
            databaseManager.save(session)
        }

        buildSession(androidFurDummies, 0.008720931f, 0.9800000190734863f, 1146, 1400856553800L).let { session ->
            androidFurDummies.sessions.add(session)
            databaseManager.save(session)
        }

        val emptyBook = Book()
        databaseManager.save(emptyBook)

        return Arrays.asList(metamorphosis, androidFurDummies, emptyBook)
    }
}